Extend compilation and adimensionalization docs (#305)
pasqal-io/qoolqit: docs/extended_usage/adimensionalization.md

Reviewer: `_0` -> `_max`

> # Adimensionalization and Compilation

Reviewer: Maybe this title should be propagated to all the doc ?

`docs/get_started/qoolqit_model.md` (twice)
`docs/index.md`
`mkdocs.yml`

Although in the last case, it becomes too large compared to the column length

This section describes how QoolQit’s dimensionless formulation connects to real physical quantities, and how compilation maps an abstract programs onto actual hardware.

> The [QoolQit Model](../get_started/qoolqit_model.ipynb) page introduces the main idea of compilation at a high level: the compiler changes the same dimensionless while keeping the physic reference physical scale used to realize it on hardware. Here, we make that idea precise by defining a reference interaction, the corresponding reference distance, and the mapping between dimensionless and physical quantities.

Reviewer 2: ```suggestion
The [QoolQit Model](../get_started/qoolqit_model.ipynb) page introduces the main idea of compilation at a high level: the compiler translates a program, defined in dimensionless units, to the physical scale used to realize it on hardware. Here, we make that idea precise by defining a reference interaction, the corresponding reference distance, and the mapping between dimensionless and physical quantities.
```

---

## Reference energy and dimensionless Hamiltonian

Every neutral-atom device is characterized by a minimum allowed atom separation $r_{\min}$, set by hardware. This minimum spacing corresponds to the largest pairwise interaction the device can ever produce:

$$
J_{\text{max}} \;=\; \frac{C_6}{r_{\min}^{6}}.
$$

QoolQit takes this $J_{\text{max}}$ as the **reference energy scale** for adimensionalization, and the corresponding minimum spacing as the reference distance.

!!! note "Key point"
    Both $r_{\text{min}}$ and $J_{\text{max}}$ are **device constants**. They are fixed by the hardware specification and do not depend on the user's program. They are not chosen by compilation.

All Hamiltonian parameters are expressed relative to this fixed scale:

$$

> \tilde{r}_{ij} = \frac{r_{ij}}{r_{\text{min}},

Reviewer 2: ```suggestion
\tilde{r}_{ij} = \frac{r_{ij}}{r_{\text{min}}},
```

\qquad

> \tilde{J}_{ij} = \frac{1}{\tilde{r}_{ij}^6} = \frac{C_6/r_{ij}^6}{J_{\text{max}}},

Reviewer: The whole equation doesn't render on my PC when built with `mkdocs`

Reviewer 2: Done!
The not rendering issue is originating from not giving a line break before the equation. Adding it render them fine:

```md
All Hamiltonian parameters are expressed relative to this fixed scale:

$$
\tilde{r}_{ij} = \frac{r_{ij}}{r_{\text{min}}},
\qquad
\tilde{J}_{ij} = \frac{1}{\tilde{r}_{ij}^6} = \frac{C_6/r_{ij}^6}{J_{\text{max}}},
$$
```

\sum_i \left(\tilde{\delta}(t) + \epsilon_i\tilde{\Delta}(t)\right)\hat{n}_i.
$$

> Since $J_0$ is the largest interaction the device can produce, every physically realizable register satisfies

Reviewer 2: ```suggestion
Since $J_{\text{max}}$ is the largest interaction the device can produce, under this renormalization, every physically realizable register satisfies:
```

Reviewer 2: renormalization or convention

$$
\min_{i<j}\tilde r_{ij} \geq 1,
\qquad\text{equivalently}\qquad
\max_{i<j}\tilde J_{ij} \leq 1,
$$

---

## From dimensionless programs to physical hardware

> A QoolQit program is specified in terms of dimensionless quantities ($\tilde{J}_{ij}$, $\tilde{\Omega}(t)$, $\tilde{\delta}(t)$, $\tilde t$). Once the device, and therefore $J_{\text{max}}$ and $r_{\text{min}}$ — is fixed, the conversion to physical units is **completely determined**:

Reviewer: The opening comma doesn't match the closing dash 
```
, and therefore ... -
```

Reviewer 2: good catch! fixed

$$
\Omega(t) = J_{\text{max}}\,\tilde{\Omega}(t),
\qquad
\delta(t) = J_{\text{max}}\,\tilde{\delta}(t),
\qquad
\Delta(t) = J_{\text{max}}\,\tilde{\Delta}(t),
\qquad
r_{ij} = r_{\text{min}}\,\tilde r_{ij}.
$$

A given dimensionless program corresponds to one and only one set of physical parameters on a given device.

In the following step, QoolQit adjusts the **dimensionless program** so that the resulting physical parameters satisfy the device's operational constraints.

---

## Compilation

> TThe user specifies a dimensionless program by providing a register (which determines the values $\tilde J_{ij}$) and a drive (which determines $\max_{\tilde t} \tilde \Omega(t)$, $\tilde\delta(t)$, etc.). Together these correspond to a point in the $(\tilde J, \tilde\Omega)$ plane — more precisely, to a value of the ratio

Reviewer 2: ```suggestion
The user specifies a dimensionless program by providing a register (which determines the values $\tilde J_{ij}$) and a drive (which determines $\max_{\tilde t} \tilde \Omega(t)$, $\tilde\delta(t)$, etc.). Together these correspond to a point in the $(\tilde J, \tilde\Omega)$ plane — more precisely, to a value of the ratio
```

$$
\frac{\max_{\tilde t}\tilde\Omega}{\max_{i<j}\tilde J_{ij}},
$$

which defines a line through the origin in that plane.

In addition to the upper bound $\tilde J_{ij} \leq 1$ inherent to the adimensionalization, the device imposes a maximum drive amplitude $\tilde\Omega_{\text{max}} = \Omega_{\max}/J_{\text{max}}$. Together these define a **device allowed region** represented as a shaded green rectangle in the figure below:

![Compilation diagram](../extras/assets/compilation.svg)

> If the user's point lies outside this region, the program cannot be implemented as specified. If it lies strictly inside, the program is feasible but does not exploit the full capability of the device. Compilation resolves both situations by sliding the program point along the line until it sits exactly on the boundary of the feasible region, maximizing $\tilde\Omega$.

Reviewer 2: ```suggestion
If the user's point lies outside this region, the program cannot be implemented as specified. If it lies strictly inside, the program is feasible but does not exploit the full capability of the device. Compilation resolves both situations by rescaling the program (sliding the point along the line) until it sits exactly on the boundary of the feasible region, maximizing $\tilde\Omega$. 
```

Reviewer 2: I would also reiterate here with something like:

> It is important to remind that this rescaling does not alter the prorgram, rather simply set the most favorable energy scale at which the program should be realized

> Concretely, compilation multiplies all dimensionless parameters along the ray by a common factor $\alpha$:

Reviewer 2: ```suggestion
Concretely, compilation rescale all dimensionless parameters by a common factor $\alpha$:
```

$$
\tilde J_{ij}\;\to\;\alpha\,\tilde J_{ij},
\qquad
\tilde\Omega\;\to\;\alpha\,\tilde\Omega,
\qquad
\tilde\delta\;\to\;\alpha\,\tilde\delta,
$$

with $\alpha$ chosen as large as possible while keeping the program inside the feasible region. The ratio $\tilde\Omega/\tilde J$ is preserved by construction, so the dimensionless content of the program — the relative balance between drive and interactions — is unchanged.

### Drive-limited compilation

> When the ratio $\tilde\Omega/\tilde J$ is large, the ray hits the line $\tilde\Omega = \tilde\Omega_{\max}$ before reaching $\tilde J = 1$. The compiled program saturates the drive maximum:

Reviewer 2: ```suggestion
When the ratio $\tilde\Omega/\tilde J$ is large, the ray hits the line $\tilde\Omega = \tilde\Omega_{\max}$ before reaching $\tilde J = 1$. The compiled program saturates the drive maximum amplitude:
```

In physical units,

$$
i\hbar \frac{d}{dt}|\psi(t)\rangle = H(t)|\psi(t)\rangle.
$$

Because $J_{\text{max}}$ is a fixed device constant, the change of variable $\tilde t = J_{\text{max}}t$ rewrites this exactly as

$$
i\frac{d}{d\tilde t}|\psi(\tilde t) \rangle = \tilde H(\tilde t)|\psi(\tilde t) \rangle,
$$

where $\tilde H = H/J_{\text{max}}$. The conversion $\tilde t = J_{\text{max}} t$ between physical and dimensionless time is therefore **fixed by the device**, exactly like $\tilde \Omega = \Omega/J_{\text{max}}$.

Now consider what happens during compilation. Compilation multiplies the dimensionless Hamiltonian by a factor $\alpha$:

$$
\tilde H(\tilde t)\;\longrightarrow\;\alpha\,\tilde H(\tilde t).
$$

For two programs along the same line to represent the **same physics**,i.e. to generate the same unitary evolution, the dimensionless time variable must be rescaled by the inverse factor:

$$
\tilde t\;\longrightarrow\;\tilde t/\alpha.
$$

Indeed, with this rescaling,

$$
\mathcal{T}\exp\left(-i\int_0^{\tilde t/\alpha}\alpha\,\tilde H(\tilde t')\,d\tilde t'\right)
=
\mathcal{T}\exp\left(-i\int_0^{\tilde t}\tilde H(s)\,ds\right),
$$

> so the unitary generated by the rescaled program over a rescaled dimensionless duration is identical to the unitary generated by the original program.

Reviewer 2: ```suggestion
the unitary generated by the rescaled program is identical to the unitary generated by the original program.
```

---

## Compiling time back to physical units

Combining the fixed device conversion $t = \tilde t/J_{\text{max}}$ with the compilation rescaling $\tilde t \to \tilde t/\alpha$, a dimensionless duration $\tilde T$ specified by the user is realized physically in a time

$$
T \;=\; \frac{\tilde T/\alpha}{J_{\text{max}}} \;=\; \frac{\tilde T}{\alpha\,J_{\text{max}}}.
$$

A larger $\alpha$, that is a program with larger amplitude and interaction, yields a shorter physical runtime. This is consistent with the compilation strategy: by maximizing $\alpha$, QoolQit produces the fastest physical implementation compatible with the device constraints.

---

## Physical interpretation of dimensionless time

> Let us definethe strongest interaction **present in the register** as

Reviewer 2: ```suggestion
Let us define the strongest interaction **present in the register** as
```

$$
\tilde J_{\text{max}} \;=\; \max_{i<j}\tilde J_{ij} \;\leq\; 1.

> $$

Reviewer 2: already defined, right?

Reviewer: Doesn't render on my side, when compiled with `mkdocs`

Author: the adimensional maximum was not yet defined

> In an interacting many-body system, the product $\tilde J_{\text{max}}\tilde t$ (and so the time $\tilde $) has a natural physical interpretation in terms of the buildup and propagation of correlations. Following the Lieb-Robinson picture, correlations spread at a finite speed set by the interaction scale. Roughly speaking:

Reviewer 2: ```suggestion
In an interacting many-body system, the product $\tilde J_{\text{max}}\tilde t$ (and so the time $t$) has a natural physical interpretation in terms of the buildup and propagation of correlations. Following the Lieb-Robinson picture, correlations spread at a finite speed set by the interaction scale. Roughly speaking:
```

- $\tilde \tilde t \ll 1/J_{\text{max}}$ corresponds to evolution that is too short for interactions to significantly affect the dynamics;
- $\tilde \tilde t \sim 1/J_{\text{max}}$ corresponds to the timescale on which nearest-neighbor correlations can begin to emerge;

> - $\tilde \tilde t \sim n/J_{\text{max}}$ can be interpreted as the timescale on which correlations may have propagated across a distance of order $n$ lattice spacings, assuming approximately ballistic spreading.

Reviewer: None renders on my side when compiled with `mkdocs`. I get the error
```
Missing argument for \tilde
```

Reviewer 2: Additional unwanted \tilde removed. Done!

> - $\tilde \tilde t \sim n/J_{\text{max}}$ can be interpreted as the timescale on which correlations may have propagated across a distance of order $n$ lattice spacings, assuming approximately ballistic spreading.

Reviewer 2: ```suggestion
- $\tilde t \ll 1/J_{\text{max}}$ corresponds to evolution that is too short for interactions to significantly affect the dynamics;
- $\tilde t \sim 1/J_{\text{max}}$ corresponds to the timescale on which nearest-neighbor correlations can begin to emerge;
- $\tilde t \sim n/J_{\text{max}}$ can be interpreted as the timescale on which correlations may have propagated across a distance of order $n$ lattice spacings, assuming approximately ballistic spreading.
```

Since, $J_{\text{max}}$ is the largest interaction the device can produce, under this renormalization, every physically realizable register satisfies:

$$
\min_{i<j}\tilde r_{ij} \geq 1,
\qquad\text{equivalently}\qquad
\max_{i<j}\tilde J_{ij} \leq 1,
$$

---

## From dimensionless programs to physical hardware

A QoolQit program is specified in terms of dimensionless quantities ($\tilde{J}_{ij}$, $\tilde{\Omega}(t)$, $\tilde{\delta}(t)$, $\tilde t$). Once the device, and therefore $J_{\text{max}}$ and $r_{\text{min}}$ are fixed, the conversion to physical units is **completely determined**:

$$
\Omega(t) = J_{\text{max}}\,\tilde{\Omega}(t),
\qquad
\delta(t) = J_{\text{max}}\,\tilde{\delta}(t),
\qquad
\Delta(t) = J_{\text{max}}\,\tilde{\Delta}(t),
\qquad
r_{ij} = r_{\text{min}}\,\tilde r_{ij},
\qquad
t = \tilde t/J_{\text{max}}.
$$

Thus, a given dimensionless program corresponds to one and only one set of physical parameters on a given device.

In the following step, QoolQit adjusts the **dimensionless program** so that the resulting physical parameters satisfy the device's operational constraints.

---

## Compilation

> The user specifies a dimensionless program by providing a register (which determines the values $\tilde J_{ij}$) and a drive (which determines $\max_{\tilde t} \tilde \Omega(t)$, $\tilde\delta(t)$, etc.). Together these correspond to a point in the $(\tilde J, \tilde\Omega)$ plane — more precisely, to a value of the ratio

Reviewer 2: Integrate better detuning here

Reviewer 2: Done by clearly specifying that we will only look at ($J$, $\Omega$) plots

\qquad
r_{ij} = r_{\text{min}}\,\tilde r_{ij},
\qquad
t = \tilde t/J_{\text{max}}.
$$

Thus, a given dimensionless program corresponds to one and only one set of physical parameters on a given device.

In the following step, QoolQit adjusts the **dimensionless program** so that the resulting physical parameters satisfy the device's operational constraints.

### Energy scale maximization

We will now discuss the compilation strategy that QoolQit uses to rescale the compiled program so that it can both run on a physical QPU and fully exploit its capabilities.

As discussed, the user specifies a dimensionless program by providing a register (which determines the values $\tilde J_{ij}$) and a time-dependent drive (amplitude, detuning and phase, which determines $\tilde \Omega(t)$, $\tilde\delta(t)$, $\tilde\Delta(t)$ and $\phi$).
Over all these parameters, two in particular are arguably more important since they set the energy scales of the interactions and of the drive, and thus their ratio.
Respectively, they are the maximum interaction $\max_{i>j}\tilde J_{ij}$ and the maximum driving amplitude $\max_{\tilde t}\tilde\Omega$ in the defined program.

> Simplifying, below, we will represent a program as a point in the $(\tilde J, \tilde\Omega)$ plane. Interestingly, every program also defines a line through the origin in that plane. Every line defines a "family" of equivalent programs, where the ratio between interaction and driving strength is fixed.

Reviewer: > "family" of equivalent programs

is never proprely/explicitely defined.

In my own words, it's that the set
$\lbrace \(\frac{H}{\alpha}, \alpha t\), \forall \alpha \in R^{+*} \rbrace$
describes the same physical system

Reviewer 2: Rephrased as

```
Thus, simplifying, we will represent a program as a point in the $(\tilde J, \tilde\Omega)$ plane. Interestingly, every program also defines a line through the origin in that plane, with slope $\tilde{\Omega}_{\text{max}}/\tilde{J}_{\text{max}}$. All points lying on the same line are ideally physically identical programs, differing only by a global energy scale factor while preserving the fundamental ratio between driving and interaction strengths.
```

!!! note "Why maximize the energy scale?"
    Higher amplitude guarantees better signal/noise ratio and thus better results, while higher interactions, thus lower distances between atoms help making the register more compact, allowing to put more qubits in your program.

Concretely, compilation rescales all dimensionless parameters by a common factor $\alpha$:

$$
\tilde J_{ij}\;\to\;\alpha\,\tilde J_{ij},
\qquad
\tilde\Omega\;\to\;\alpha\,\tilde\Omega,
\qquad
\tilde\delta\;\to\;\alpha\,\tilde\delta,
$$

with $\alpha$ chosen as large as possible while keeping the program inside the feasible region. The ratio $\tilde\Omega/\tilde J$ is preserved by construction, so the dimensionless content of the program — the relative balance between drive and interactions — is unchanged.

### Drive-limited compilation

> When the ratio $\tilde\Omega/\tilde J$ is large, the ray hits the line $\tilde\Omega = \tilde\Omega_{\max}$ before reaching $\tilde J = 1$. The compiled program saturates the drive maximum amplitude:

Reviewer: ```suggestion
When the ratio $\tilde\Omega/\tilde J$ is large, the ray hits the line $\tilde\Omega = \tilde\Omega_{\max}$ before reaching $\tilde J = 1$ (blue line above). The compiled program saturates the drive maximum amplitude:
```

$$
\alpha \,\max_{\tilde t}\tilde\Omega \;=\; \tilde\Omega_{\max}
\qquad\Longrightarrow\qquad
\alpha \;=\; \frac{\tilde\Omega_{\max}}{\max_{\tilde t}\tilde\Omega}.
$$

Atoms are placed further apart than the device minimum: the closest pair sits at a dimensionless distance $\tilde r > 1$, equivalently at a physical distance $r > r_{\min}$.

### Interaction-limited compilation

> When the ratio $\tilde\Omega/\tilde J$ is small, the ray hits $\tilde J = 1$ before reaching $\tilde\Omega_{\max}$. The compiled program saturates the interaction bound:

Reviewer: ```suggestion
When the ratio $\tilde\Omega/\tilde J$ is small, the ray hits $\tilde J = 1$ before reaching $\tilde\Omega_{\max}$ (red line above). The compiled program saturates the interaction bound:
```

![Compilation diagram](../extras/assets/compilation.svg)

If the user's point lies outside this region, the program cannot be implemented as specified. If it lies strictly inside, the program is feasible but does not exploit the full capability of the device. Compilation resolves both situations by rescaling the program (sliding the point along the line) until it sits exactly on the boundary of the feasible region, maximizing $\tilde\Omega$. 

> Concretely, compilation rescale all dimensionless parameters by a common factor $\alpha$:

Reviewer: rescales

## Physical interpretation of dimensionless time

Let us define the strongest interaction **present in the register** as
$$
\tilde J_{\text{max}} \;=\; \max_{i<j}\tilde J_{ij} \;\leq\; 1.
$$

In an interacting many-body system, the product $\tilde J_{\text{max}}\tilde t$ (and so the time $\tilde $) has a natural physical interpretation in terms of the buildup and propagation of correlations. Following the Lieb-Robinson picture, correlations spread at a finite speed set by the interaction scale. Roughly speaking:

- $\tilde t \ll 1/J_{\text{max}}$ corresponds to evolution that is too short for interactions to significantly affect the dynamics;
- $\tilde t \sim 1/J_{\text{max}}$ corresponds to the timescale on which nearest-neighbor correlations can begin to emerge;

> - $\tilde t \sim n/J_{\text{max}}$ can be interpreted as the timescale on which correlations may have propagated across a distance of order $n$ lattice spacings, assuming approximately ballistic spreading.

Reviewer: Why divided by Jmax ? This is not consistent with the version in `QoolQit Model`.
https://pasqal-io.github.io/qoolqit/latest/get_started/qoolqit_model/#time-regimes

Author: you are right, here I meant divided by \tilde{J}_{\max}

Reviewer 2: all relative to $\tilde{J}_{\max}$
